### PR TITLE
health: shutdown server to set NOT_SERVING and disallow future updates

### PR DESCRIPTION
**Pull request by menghanl**

### CODE REVIEW

File: health/server.go [hunk 1]
```diff
@@ -110,7 +117,14 @@ func (s *Server) Watch(in *healthpb.HealthCheckRequest, stream healthgrpc.Health
 func (s *Server) SetServingStatus(service string, servingStatus healthpb.HealthCheckResponse_ServingStatus) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
+	if s.shutdownEvent.HasFired() {
```
Comment (dfawley) [inline health/server.go]: I think this could go above the lock.  (?)

File: health/server_internal_test.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+)
+
+func TestShutdown(t *testing.T) {
+	const testService = "tteesstt"
+	s := NewServer()
+	s.SetServingStatus(testService, healthpb.HealthCheckResponse_SERVING)
+
+	s.mu.Lock()
```
Comment (dfawley) [inline health/server_internal_test.go]: There is nothing asynchronous happening, so I don't think this needs a lock.  Remove the `status` local entirely maybe?

File: health/server.go (hunk 1 above)
Comment (menghanl) [reply health/server.go]: There can be a race if this is not in the lock

goroutine-1: `HasFired()=false`
goroutine-2: Fire()
goroutine-2: take lock, set Not_serving, release lock
goroutine-1: take lock, set status (to Serving)

File: health/server_internal_test.go (hunk 2 above)
Comment (menghanl) [reply health/server_internal_test.go]: done

**Pull request merged**